syntax = "proto3";

package protocol;

import "common.proto";

option java_package = "com.nautilus_technologies.tsubakuro.protos";
option java_outer_classname = "ResponseProtos";

/*
 * Definition of sub fields for Response.
 */

/* For response of success when there is no data to return. */
message Success {
}

/* For response of error containing a error message. */
message Error {
  string detail = 1;
}


/*
 * Each response message
 */

/* For response to ExecuteStatement, ExecutePreparedStatement, Commit, and/or Rollback. */
message ResultOnly {
  oneof result {
    Success success = 1;
    Error error = 2;
  }
}

/* For response to Begin. */
message Begin {
  oneof result {
    common.Transaction transaction_handle = 1;
    Error error = 2;
  }
}

/* For response to Prepare. */
message Prepare {
  oneof result {
    common.PreparedStatement prepared_statement_handle = 1;
    Error error = 2;
  }
}

/* For response to ExecuteQuery and/or ExecutePreparedQuery. */
message ExecuteQuery {
  oneof result {
    string name = 1;  /* The name of the channel to which the ResultSet set will be sent. */
    Error error = 2;
  }
}


/* For response message from the SQL service. */
message Response {
  oneof response {
    ResultOnly result_only = 1;
    Begin begin = 2;
    Prepare prepare = 3;
    ExecuteQuery execute_query = 4;
  }
}
